### ISSUE TITLE
Improve documentation for Eclipse IDE and maven setup

### ISSUE THREAD

**Issue by triller-telekom** [lastnpe/eclipse-null-eea-augments#52]
The Maven usage documentation is a bit unclear to me:

In https://github.com/lastnpe/eclipse-null-eea-augments/issues/29#issuecomment-342532024 @vorburger mentions that one can use the eeas as maven dependencies like this:

```
<dependencies>
<dependency>
      <groupId>org.lastnpe.eea</groupId>
      <artifactId>jdk-eea</artifactId>
      <version>0.0.1-SNAPSHOT</version>
      <scope>provided</scope>
    </dependency>
```

and then configure the jdt compiler like this

```
<compilerId>jdt</compilerId>
          <compilerArguments>
            <annotationpath>CLASSPATH</annotationpath>
```

But WHO sets this "CLASSPATH"? Is it set by maven and filled with everything from the "<dependencies>" section?

Regarding the use within Eclipse:

You are saying that https://github.com/lastnpe/eclipse-external-annotations-m2e-plugin can be used to setup my projects to use the jars inside a maven repo. But how?

> Allows to configure Java Compiler Project Properties from maven-compiler-plugin; read either from a dependency of maven-compiler-plugin containing a org.eclipse.jdt.core.prefs file, or from configuration/compilerArguments/properties.

Here I understand that I can put a "org.eclipse.jdt.core.prefs" config file in my repository that has setting for eea. But how and when ware they applied to my projects within Eclipse?

> Allows to configure the path to external annotations for the Maven Dependencies and JRE classpath containers. The path is either taken from the "m2e.jdt.annotationpath" property in POM files, if it exists (this for configuring one single archive of *.eea for ALL Maven dependencies AND the JRE),

From this I understand that I can set the "m2e.jdt.annotationpath" property in my pom.xml and whatever is in there will be copied into my project settings for all my projects in the workspace? When and how? Also what do I write into that property? "CLASSPATH"? like I do for the "annotationpath" in the jdt settings of the pom.xml?

Sorry about these questions, but I am still learning maven and that is why I am struggeling to understand the examples you are providing. Maybe this way the documentation can be improved...

**Comment by vorburger** [comment]
> But WHO sets this "CLASSPATH"? Is it set by maven and filled with everything from the "" section?

does the new inline doc proposed to be added in https://github.com/lastnpe/eclipse-null-eea-augments/pull/53 help clarify how that magic works?

> eclipse-external-annotations-m2e-plugin can be used to setup my projects to use the jars inside a maven repo. But how?

Have you tried out to just install that Eclipse plugin and open the examples/maven/ projects?  Seeing it in action may be the best way to learn about it... :smiley: 

Yes you [put a "org.eclipse.jdt.core.prefs" config file into a Maven artifact](https://github.com/vorburger/eclipse-null-eea-augments/tree/master/examples/maven/jdt-ecj-settings), and then [have a dependency to it](https://github.com/vorburger/eclipse-null-eea-augments/blob/4185c9e0d0ed5cb38aeb06fae933d70137263dc6/examples/maven/parent/pom.xml#L77), and [eclipse-external-annotations-m2e-plugin will find and set it on Maven project configuration](https://github.com/lastnpe/eclipse-external-annotations-m2e-plugin/blob/master/eclipse-external-annotations-m2e-plugin.core/src/org/lastnpe/m2e/core/configurator/ClasspathConfigurator.java#L314).

> setting for eea. But how and when ware they applied to my projects within Eclipse?

Basically the eclipse-external-annotations-m2e-plugin extends the M2E Maven Eclipse integration, and that Eclipse plugin calls this Eclipse plugin whenever you e.g. right-click Project > Maven > Update Project. You can learn more about it in its source code! :smiling_imp: 

The eclipse-external-annotations-m2e-plugin really has two completely separate functionalities, and you are welcome to use only one or ther other of them (or neither); whatever is useful to you.

Hope this helps? Let me know when you think this issue can be closed. Or feel free to raise a PR contributing better documentation! :smile_cat:

**Comment by triller-telekom** [comment]
> does the new inline doc proposed to be added in #53 help clarify how that magic works?

Thank you for adding this, yes it does help and I managed to get it to work with our tycho-compiler plugin by specifying an `extraClasspathElement` like this:

```
<plugin>
          <groupId>${tycho-groupid}</groupId>
          <artifactId>tycho-compiler-plugin</artifactId>
          <version>${tycho-version}</version>
          <configuration>
            <extraClasspathElements>
              <extraClasspathElement>
                <groupId>org.apache.felix</groupId>
                <artifactId>org.apache.felix.scr.ds-annotations</artifactId>
                <version>${ds-annotations.version}</version>
              </extraClasspathElement>
              <extraClasspathElement>
                <groupId>org.eclipse.jdt</groupId>
                <artifactId>org.eclipse.jdt.annotation</artifactId>
                <version>${jdt-annotations.version}</version>
              </extraClasspathElement>
              <extraClasspathElement>
                <!-- TODO: insert one project that contains ALL annotations here-->
                <groupId>org.lastnpe.eea</groupId>
                <artifactId>jdk-eea</artifactId>
                <version>0.0.1</version>
              </extraClasspathElement>
            </extraClasspathElements>
            <compilerArgs>
              <arg>-err:+nullAnnot(org.eclipse.jdt.annotation.Nullable|org.eclipse.jdt.annotation.NonNull|org.eclipse.jdt.annotation.NonNullByDefault),+inheritNullAnnot</arg>
              <arg>-warn:+null,+inheritNullAnnot,+nullAnnotConflict,+nullUncheckedConversion,+nullAnnotRedundant,+nullDereference</arg>
              <arg>-annotationpath</arg>
              <arg>CLASSPATH</arg>
            </compilerArgs>
          </configuration>
        </plugin>
```

Regarding the "try out and test" of your examples within Eclipse. I can import the projects if I select "Import Maven Project" so I get a bunch of maven projects in the IDE.

The problem I have in my ESH IDE is that our projects are "Java Projects", not "Maven projects", so I do not have "right-click Project > Maven > Update Project". So I guess we cannot make use of it?

Is there another way of telling the Eclipse IDE that it should put a bunch of maven dependencies onto the Classpath and use it when resolving the external annotations? In the external annotations dialog I can only select paths and if I try to enter the "CLASSPATH" it says that the directory does not exist.

Thank you for your help so far!

**Comment by vorburger** [comment]
> managed to get it to work with our tycho-compiler plugin by specifying an extraClasspathElement

Cool. In an ideal world, if you have the time, then a contribution with that in an `example/tycho` to this repo could be neat, for others to find in the future. Or they'll just find this issue by searching, that's fine.

> The problem I have in my ESH IDE is that our projects are "Java Projects", not "Maven projects", so I do not have "right-click Project > Maven > Update Project". So I guess we cannot make use of it?

Oh! Yeah... the lastnpe/eclipse-external-annotations-m2e-plugin is (quote) a _"Eclipse Maven integration (M2E extension) for null analysis"_ so... yeah, if you configure your projects differently, then you need to figure out a solution within the toolset you guys in ESH use!

Hope this helps? Let me know when you think this issue can be closed.

**Comment by triller-telekom** [comment]
I have just created PR #54 with an example on how to configure the tycho compiler.

Regarding the other issue with the IDE we are now using the "maven-dependency-plugin" to download the eea artifact from maven central and copy the containing jar file into a directory that can be accessed by the IDE. This maven task is called as a setup task from our ESH IDE setup.

We can close this issue now. Thank you for your help!

The next steps will be to work on #51 and get a single artifact that contains all annotations.
